jhdf/src/main/java/io/jhdf/storage/HdfFileChannel.java

```diff
@@ -73,7 +73,35 @@ public ByteBuffer map(long address, long length) {
 	@Override
 	public ByteBuffer mapNoOffset(long address, long length) {
 		try {
+			try {
 			return fc.map(MapMode.READ_ONLY, address, length);
+			} catch (UnsupportedOperationException e) {
+				// many file systems do not support memory mapping
+				int lengthAsInt = (int) length;
+				if (lengthAsInt != length) {
+					// length must be representable by an int such that we can allocate a ByteBuffer manually
+					throw e;
+				}
+				ByteBuffer buffer = ByteBuffer.allocate(lengthAsInt);
+				long oldPosition = fc.position();
+				fc.position(address);
+				try {
+					int totalNumberOfBytesRead = 0;
+					while (totalNumberOfBytesRead < lengthAsInt) {
+						int numberOfBytesRead = fc.read(buffer);
+						if (numberOfBytesRead < 0) {
+							throw new IOException("Trying to read more bytes than available");
+						}
+						totalNumberOfBytesRead += numberOfBytesRead;
+					}
+				}
+				finally {
+					// reset file channel to old position
+					fc.position(oldPosition);
+				}
+				buffer.flip();
+				return buffer;
```

Comment:
Think maybe this is a cleaner impl
```
ByteBuffer buffer = ByteBuffer.allocate(Math.toIntExact(length));
int bytesRead = fc.read(buffer, address);
if(bytesRead != length) {
	throw new IOException("Incorrect bytes read");
}
buffer.flip();
return buffer;
```
What do you think?

Reply:
Regarding Math.toIntExact(length): If this conversion fails, then the caller will get an ArithmeticException. This might be a bit surprising, so I tried to handle this case individually. However, since HdfInMemoryByteBuffer.mapNoOffset() also uses this method, I think that it is consistent to use it here as well.

Regarding your suggestion to remove the loop for reading data into the ByteBuffer: FileChannel.read() does not guarantee to fill the whole buffer. It might even return 0 although there is still more data to read. Maybe this case occurs when data is transferred via the network and it is known that there is more data, but it has not arrived yet. This is why I think it is safer to read the data in a loop.

Regarding resetting the position: I tried to avoid side effects, so I decided to reset the position after reading the data. If you think that this side effect is uncritical, then we could remove this behavior to make the code simpler. What do you think?